crate/operator/restore_backup.py

```diff
@@ -0,0 +1,925 @@
+    async def handle(  # type: ignore
+        self,
+        namespace: str,
+        name: str,
+        repository: str,
+        snapshot: str,
+        tables: List,
+        logger: logging.Logger,
+        **kwargs: Any,
+    ):
+        async with ApiClient() as api_client:
+            core = CoreV1Api(api_client)
+            data = await get_source_backup_repository_data(
+                core,
+                namespace,
+                name,
+                logger,
+            )
+            password, host = await asyncio.gather(
+                get_system_user_password(core, namespace, name),
+                get_host(core, namespace, name),
+            )
+            conn_factory = connection_factory(host, password)
+
+            await self._create_backup_repository(conn_factory, repository, data, logger)
+
+            await self._ensure_snapshot_exists(
+                conn_factory, repository, snapshot, logger
+            )
+
+            await self._start_restore_snapshot(
+                conn_factory, repository, snapshot, tables, logger
+            )
+
+    @staticmethod
+    async def _create_backup_repository(
+        conn_factory,
+        repository: str,
+        data: dict,
+        logger: logging.Logger,
+    ):
+        """
+        Create a backup repository with the given credentials if it does not exist yet.
+
+        :param conn_factory: A function that establishes a database connection to
+            the CrateDB cluster used for SQL queries.
+        :param repository: The name of the repository to be created.
+        :param data: a dict containing the bucket name, base path and secrets to access
+            the source backup repository.
+        :param logger: the logger on which we're logging
+        """
+        try:
+            async with conn_factory() as conn:
+                async with conn.cursor() as cursor:
+                    await cursor.execute(
+                        "SELECT * FROM sys.repositories WHERE name=%s", (repository,)
+                    )
+                    row = await cursor.fetchone()
+                    if not row:
+                        repository_ident = quote_ident(repository, cursor._impl)
+                        try:
+                            await cursor.execute(
+                                f"CREATE REPOSITORY {repository_ident} type s3 with "
+                                "(access_key = %s, secret_key = %s, bucket = %s, "
+                                "base_path = %s, readonly=true)",
+                                (
+                                    data["accessKeyId"],
+                                    data["secretAccessKey"],
+                                    data["bucket"],
+                                    data["basePath"],
+                                ),
+                            )
+                        except ProgrammingError as e:
+                            logger.warning(
```

Comment:
When can this happen? Should this also raise `PermanentError`? Why not add it to the exception handler below?

Reply:
true, this actually inherits from `DatabaseError` and can be handled below..